### DISCUSSION TITLE
Anyway way to print debug output from your external scanner?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1218] [Q&A] [answered]
Would be nice to see some custom output from the external scanner itself when running `tree-sitter parse -d`. Any way to do it?

**Comment by razzeee** [comment]
Yes, you can just use `sprintf("Your log here.")`. The tricky part is, that it will show up at the end and not chronologically.

If you add a newline, it will be in the correct places. `sprintf("Your log here.\n")`

**Reply by dcreager** [reply]
I think you want `printf("Your log here.\n")` or `fprintf(stderr, "Your log here.\n")` depending on whether you want your log messages to go to stdout or stderr.  (`sprintf` writes into a string in memory.)

**Comment by ahlinc** [comment]
You can get attention to this PR: https://github.com/tree-sitter/tree-sitter/pull/1187

**Comment by ahlinc** [comment, marked as answer]
@ahelwer If you want to depend on the `-d` flag for the `tree-sitter parse` command then #1186 is what you're looking for. And also I'm waiting for #1187 to make it even more agile and make it possible to compile printing logic in external scanners only if those are used by tree-sitter CLI commands.

/cc: @maxbrunsfeld

**Reply by ahlinc** [reply]
@ahelwer Now #1187 and #1186 are merged and detailed examples are there in those PRs.

**Reply by mikehaertl** [reply]
@ahlinc Forgive the stupid question, but could you maybe also give an example for how to use this from C? I'm not a super C expert but I'd expect this to work but it doesn't:

```c
#include <stdio.h>
bool tree_sitter_phpdoc_external_scanner_scan(void *payload, TSLexer *lexer,
                                                  const bool *valid_symbols) {
  if (getenv("TREE_SITTER_DEBUG")) {
    printf("SCANNER DEBUG\n");
  }
  printf("I can see this\n");

 // ...
}
```

**Reply by ahlinc** [reply]
@mikehaertl I've checked your code and it's correct, to see the output you need to call `tree-sitter parse` with `-d` flag, in that way tree-sitter will set the env variable for its parse command. Could you share a link on a git branch with the problem code? Or showing what you'd get for a command like below:
```
tree-sitter parse -d <FILE> |& grep 'SCANNER DEBUG'
```

**Reply by mikehaertl** [reply]
Nevermind. The parser I work on still uses fixed version 0.17.3. It works with a current version.

One thing: It seems, the env var is not set for `tree-sitter test -d`, right? I'd find this very helpful, though. What do you think?

**Reply by danielo515** [reply]
Is this exceptionally nice feature undocumented?

**Reply by leotrs** [reply]
I do believe it is undocumented, but it just saved my life <3

**Reply by ahlinc** [reply]
@mikehaertl the env var for the test subcommand was added in #2020
